File: benchmarks/python/list_ops.py
"""List Operations Benchmark for Python"""

def make_shuffled(n):
    return [(i * 1103515245 + 12345) % n for i in range(n)]

def bench_append(n, iters):
    checksum = 0
    for _ in range(iters):
        lst = []
        for i in range(n):
            lst.append(i)
        checksum += len(lst)
    return checksum

def bench_sort(shuffled, iters):
    checksum = 0
    for _ in range(iters):
        s = sorted(shuffled)
        checksum += len(s) + s[0] + s[len(s) // 2] + s[-1]
    return checksum

def bench_filter(lst, iters):
    checksum = 0
    for _ in range(iters):
        evens = [x for x in lst if x % 2 == 0]
        checksum += len(evens)
    return checksum

def bench_fold(lst, iters):
    checksum = 0
    for _ in range(iters):
        total = 0
        for x in lst:
            total += x
        checksum += total
    return checksum

def bench_reverse(lst, iters):
    checksum = 0
    for _ in range(iters):
        rev = lst[::-1]
        checksum += len(rev)
    return checksum

def bench_concat(a, b, iters):
    checksum = 0
    for _ in range(iters):
        combined = a + b
        checksum += len(combined)
    return checksum

if __name__ == "__main__":
    N = 10000
    append_check = bench_append(N, 500)
    print(f"append checksum: {append_check}")
    shuffled = make_shuffled(N)
    sort_check = bench_sort(shuffled, 200)
    print(f"sort checksum: {sort_check}")
    seq = list(range(N))
    filter_check = bench_filter(seq, 500)
    print(f"filter checksum: {filter_check}")
    fold_check = bench_fold(seq, 2000)
    print(f"fold checksum: {fold_check}")
    reverse_check = bench_reverse(seq, 1000)
    print(f"reverse checksum: {reverse_check}")
    half = N // 2
    a = list(range(half))
    b = list(range(half, N))
    concat_check = bench_concat(a, b, 1000)
    print(f"concat checksum: {concat_check}")

File: benchmarks/go/list_ops.go
package main

import (
	"cmp"
	"fmt"
	"slices"
)

func makeShuffled(n int) []int {
	list := make([]int, n)
	for i := 0; i < n; i++ {
		list[i] = (i*1103515245 + 12345) % n
	}
	return list
}

func benchAppend(n int, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		list := make([]int, 0)
		for i := 0; i < n; i++ {
			list = append(list, i)
		}
		checksum += len(list)
	}
	return checksum
}

func benchSort(shuffled []int, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		cpy := make([]int, len(shuffled))
		copy(cpy, shuffled)
		// Blorp List.sort is stable, so use Go's stable sorter for parity.
		slices.SortStableFunc(cpy, cmp.Compare[int])
		checksum += len(cpy) + cpy[0] + cpy[len(cpy)/2] + cpy[len(cpy)-1]
	}
	return checksum
}

func benchFilter(list []int, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		evens := make([]int, 0)
		for _, x := range list {
			if x%2 == 0 {
				evens = append(evens, x)
			}
		}
		checksum += len(evens)
	}
	return checksum
}

func benchFold(list []int, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		total := 0
		for _, x := range list {
			total += x
		}
		checksum += total
	}
	return checksum
}

func benchReverse(list []int, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		rev := make([]int, len(list))
		n := len(list)
		for i := 0; i < n; i++ {
			rev[i] = list[n-1-i]
		}
		checksum += len(rev)
	}
	return checksum
}

func benchConcat(a []int, b []int, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		combined := make([]int, 0, len(a)+len(b))
		combined = append(combined, a...)
		combined = append(combined, b...)
		checksum += len(combined)
	}
	return checksum
}

func main() {
	const N = 10000

	appendCheck := benchAppend(N, 500)
	fmt.Printf("append checksum: %d\n", appendCheck)

	shuffled := makeShuffled(N)
	sortCheck := benchSort(shuffled, 200)
	fmt.Printf("sort checksum: %d\n", sortCheck)

	// Build a sequential list for filter/fold/reverse
	seq := make([]int, N)
	for i := 0; i < N; i++ {
		seq[i] = i
	}

	filterCheck := benchFilter(seq, 500)
	fmt.Printf("filter checksum: %d\n", filterCheck)

	foldCheck := benchFold(seq, 2000)
	fmt.Printf("fold checksum: %d\n", foldCheck)

	reverseCheck := benchReverse(seq, 1000)
	fmt.Printf("reverse checksum: %d\n", reverseCheck)

	// Split for concat benchmark
	half := N / 2
	a := seq[:half]
	b := seq[half:]

	concatCheck := benchConcat(a, b, 1000)
	fmt.Printf("concat checksum: %d\n", concatCheck)
}
